x-pack/plugins/files/server/plugin.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  PluginInitializerContext,
+  CoreSetup,
+  Plugin,
+  Logger,
+  CoreStart,
+} from '@kbn/core/server';
+
+import { PLUGIN_ID } from '../common/constants';
+
+import { BlobStorageService } from './blob_storage_service';
+import { FileServiceFactory } from './file_service';
+import type { FilesPluginSetupDependencies, FilesSetup, FilesStart } from './types';
+import { fileKindsRegistry } from './file_kinds_registry';
+import type { FilesRequestHandlerContext, FilesRouter } from './routes/types';
+import { registerRoutes } from './routes';
+
+export class FilesPlugin implements Plugin<FilesSetup, FilesStart, FilesPluginSetupDependencies> {
+  private readonly logger: Logger;
+  private fileServiceFactory: undefined | FileServiceFactory;
+  private securitySetup: FilesPluginSetupDependencies['security'];
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.logger = initializerContext.logger.get();
+  }
+
+  public setup(core: CoreSetup, deps: FilesPluginSetupDependencies): FilesSetup {
+    FileServiceFactory.setup(core.savedObjects);
+    this.securitySetup = deps.security;
+
+    core.http.registerRouteHandlerContext<FilesRequestHandlerContext, typeof PLUGIN_ID>(
+      PLUGIN_ID,
+      async (ctx, req) => {
+        return {
+          fileService: {
+            asCurrentUser: () => this.fileServiceFactory!.asScoped(req),
+            asInternalUser: () => this.fileServiceFactory!.asInternal(),
+            logger: this.logger.get('files-routes'),
+          },
+        };
+      }
+    );
+
+    const router: FilesRouter = core.http.createRouter();
+
+    core
+      .getStartServices()
+      .then(() => {
+        // File routes can only be registered during start phase because we need
+        // to give plugins a chance to register file kinds.
```

Comment:
Does this add any risk of not being supported by Core at some point? I seem to remember a document somewhere saying we should register routes in `setup`.

Reply:
That's a good question! Hopefully a reviewer from core can weigh in.

But I'd expect registering routes "late" should always be OK?

Reply:
I think you should be able to use path variables `/api/files/{type}/{id}` so that you don't have to know the file types before route registration. You could still use the file registry in the route handlers to e.g. return an error code if there's an unknown type being requested.

In terms of your current implementation. Plugin lifecycle introduce specific points in time so that we don't rely on complex ordering guarantees for certain behaviour. One such a behaviour is return codes from the server. If the Kibana HTTP server is started before the route is registered there's a brief moment of time where the route would return 404. If 404 has semantic meaning to a client, they could interpret it as "the file doesn't exist".

By requiring route registration in `setup` we know that when we start listening to HTTP requests in `start` all routes are guaranteed to have been registered.

We start all plugins before we start the http server so probably it would be fine if route registration happened synchronously during start, we would just have to "close" the API after running all plugins' start to avoid this timing behaviour.

But when you're using `getStartServices` as a notification that start has been run I'm not entirely sure when that would fire because it's based on promises and observables 🤷‍♂️

Reply:
That's a good point @rudolf , thanks for clarifying.

The reason we want to register routes after setup phase (during start) is because we want plugins to register access control tags against routes (tags: `access:myPlugin`) as well as explicitly choose which HTTP routes they want "enabled".

My understanding with `tags` is that we can only do this when registering the route? This is also why we don't parameterize `/api/files/{fileKind}/{otherValue}` and statically register `/api/files/my-file-kind/{otherValue}`.

To do that, we are exposing a simplified API for them to specify which routes should exist for which file kinds. The primary concern being leveraging security access control tags. Is there another way we could achieve this?

Reply:
I suppose we could register the routes as the file kind is registered to ensure registration happens at setup time.

Reply:
So `fileKindRegistry.register(...)` would also call `router[method]` under the hood.

Reply:
yeah doing everything synchronously when the file kind is registered should solve the problem. The only alternative would be to ask security to expose authorization api's instead of only exposing it through route tags.

Reply:
https://github.com/elastic/kibana/pull/138392